### DISCUSSION TITLE
Prop drilling v.s. using more ref.watch is preferred?

### DISCUSSION THREAD

**Discussion by kenNg1** [rrousselGit/riverpod#1780] [Q&A] [answered]
#### In general, should we try to reduce the number of `ref.watch` we have across our app/widgets by passing more arguments in between our widgets?

Using more prop drilling:

```
class ChatSidePanelContainer extends StatelessWidget {
  final List<ChatListItem>? activeChats;
  final int? selectedRoomId;

  const ChatSidePanelContainer({
    Key? key,
    this.activeChats,
    this.selectedRoomId,
  }) : super(key: key);

   @override
  Widget build(BuildContext context) {
```

Using more `ref.watch`
```
class ChatSidePanelContainer extends HookConsumerWidget {
  const ChatSidePanelContainer({Key? key}) : super(key: key);
 
   @override
  Widget build(BuildContext context, WidgetRef ref) {
    final chatDetailNotifier = ref.watch(chatDetailProvider.notifier);

    final selectedRoomId =
        ref.watch(chatDetailProvider.select((value) => value.chatDetail?.roomId));

    final activeChats =
        ref.watch(allChatActiveProvider.select((value) => value.activeChats));

```

The above Sample Code is a bit contrived, but I feel it could be relevant if you have a hierarchy such as:

Widget A (parent of B) - uses chatDetailProvider
Widget B (parent of C) - doesn't use chatDetailProvider
Widget C - uses chatDetailProvider

#### Some of the potential pros/cons I can think of are listed below...

Pros:
- It slightly easier to unit test, as the widgets are essentially 'dumb' widgets. 
- Potentially more performant as we have fewer observers doing `ref.watch`

Cons:
- Implementation code is slightly more verbose, as we need to declare the fields in the widget constructor

I personally don't know which is best - I'd like to know from the community if you feel the pros and cons I've listed are valid?

**Comment by rrousselGit** [comment, marked as answer]
It's kind of your call. There's no finite answer

Although one thing worth noting with the `watch` variant is: It can now use `const` constructors. So the widget would not rebuild even if its parent did.
That can be a significant benefit.
